### ISSUE TITLE
Fix Ruby parse errors in QA repositories

### ISSUE THREAD

**Issue by mschwager** [semgrep/semgrep#1457]
The following Ruby projects are producing Ruby parse errors:

* https://github.com/rails/rails
* https://github.com/rapid7/metasploit-framework
* https://github.com/Homebrew/brew
* https://github.com/gitlabhq/gitlabhq

The parse errors can be found here: https://github.com/returntocorp/semgrep/pull/1456/checks?check_run_id=950616441

Let's fix these up and remove their `xfail_repo` designation :+1:

**Comment by mschwager** [comment]
cc @aryx @mjambon

**Comment by nbrahms** [comment]
As a first pass, we need to identify code snippets that fail and break them down into further issues. @mschwager are you able to help out with this?

**Comment by mschwager** [comment]
> As a first pass, we need to identify code snippets that fail and break them down into further issues. @mschwager are you able to help out with this?

I did a quick smoketest against `rails`:

```
$ semgrep-core -lang ruby -test_parse_lang rails/
242 / 2812/tmp/rails/actionpack/lib/action_controller/metal.rb: exn = Parse_info.Parsing_error(_)
340 / 2812/tmp/rails/actionpack/lib/action_dispatch/routing/mapper.rb: exn = Parse_info.Parsing_error(_)
355 / 2812/tmp/rails/actionpack/lib/action_dispatch/testing/assertions/routing.rb: exn = Parse_info.Parsing_error(_)
768 / 2812/tmp/rails/actionview/test/template/form_helper/form_with_test.rb: exn = Parse_info.Parsing_error(_)
769 / 2812/tmp/rails/actionview/test/template/form_helper_test.rb: exn = Parse_info.Parsing_error(_)
1238 / 2812/tmp/rails/activerecord/lib/active_record/relation/query_methods.rb: exn = Parse_info.Parsing_error(_)
1294 / 2812/tmp/rails/activerecord/lib/arel/attributes/attribute.rb: exn = Parse_info.Parsing_error(_)
1360 / 2812/tmp/rails/activerecord/lib/arel/visitors/dot.rb: exn = Parse_info.Parsing_error(_)
1502 / 2812/tmp/rails/activerecord/test/cases/arel/select_manager_test.rb: exn = Parse_info.Parsing_error(_)
2812 / 2812---------------------------------------------------------------
NB total files = 2812; NB total lines = 400084; perfect = 2803; pbs = 9; timeout = 0; =========> 99%
nb good = 388179,  nb passed = 0 =========> 0.000000%
nb good = 388179,  nb bad = 11905 =========> 97.024375%
```

If we need more granular parsing error information, i.e. what language feature is causing the issue, this will definitely take more time. Looking at the CI failure above, it looks like there's around 100 files that fail to parse across all the repos. I may be able to look into this more closely next week, but it'd probably be quicker for @aryx or @mjambon to take a look at `semgrep-core` running against the repos above.

**Comment by brendongo** [comment]
Filed #1580 #1581 #1582 #1583 for some errors out of rails project. It's weird in that the parse errors that are being reported don't actually parse incorrectly; there seems to be an underlying issue that then reports on these lines. Stopping splitting into separate issue hoping that the root cause of the 4 above remove a lot of other issues with these repos.

**Comment by brendongo** [comment]
Marking as blocked for now. Should revisit after the above issues are looked at.

**Comment by aryx** [comment]
Closed 3 out of the 4 submitted by brendon

**Comment by aryx** [comment]
rails now parse correctly at 100% (not sure why the xfail in our tests is not a PASS now).

**Comment by aryx** [comment]
gitlabhq has just 1 parse error! (on 15000 files)

**Comment by aryx** [comment]
metasploit has 7 parse errors, so 99.94% correct parsing

**Comment by aryx** [comment]
and 1 parse error in brew! 99.89% success

**Comment by DrewDennison** [comment]
@mschwager lots of progress here! Can you break out any remaining failing parsing errors into separate tickets so we can close this?

**Comment by mschwager** [comment]
> @mschwager lots of progress here! Can you break out any remaining failing parsing errors into separate tickets so we can close this?

Looks like we can fix this by moving `rails` from `xfail_repo`, and we'll have to wait on the others. The only remaining ticket is https://github.com/returntocorp/semgrep/issues/1580.

**Comment by mschwager** [comment]
Looks like https://github.com/returntocorp/semgrep/issues/2097 is blocking `rails` as a QA repo.

**Comment by mschwager** [comment]
Closing out with addition of `rails` in https://github.com/returntocorp/semgrep/pull/2257.
